### ISSUE TITLE
ConcurrentModificationException in Optimizely.filterAttributes()

### ISSUE THREAD

**Issue by fingertricks** [optimizely/android-sdk#227]
We are seeing a low-frequency crash originating within the Optimizely Android SDK (v2.1.0). This is observed when calling OptimizelyClient.isFeatureEnabled():

```
Caused by java.util.ConcurrentModificationException
       at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:411)
       at java.util.LinkedHashMap$EntryIterator.next(LinkedHashMap.java:430)
       at java.util.LinkedHashMap$EntryIterator.next(LinkedHashMap.java:430)
       at com.optimizely.ab.Optimizely.filterAttributes(Optimizely.java:776)
       at com.optimizely.ab.Optimizely.isFeatureEnabled(Optimizely.java:362)
       at com.optimizely.ab.android.sdk.OptimizelyClient.isFeatureEnabled(OptimizelyClient.java:360)
...
```

As a temporary workaround we have added a try...catch around calls to ```isFeatureEnabled()```.

We've seen around 20 crashes in the last couple of weeks since we integrated the Optimizely SDK.

**Comment by fingertricks** [comment]
After further investigation, we've found the cause of this that we were passing a mutable attributes map to Optimizely, and in rare cases that map could be modified on a background thread. We've resolved this by passing an immutable map of attributes into activate() or isFeatureEnabled().

Leaving this open in case the SDK team want to consider copying the attributes passed in, or making the documentation clearer.
